pkg/azuredisk/azure_controller_common.go

```diff
@@ -190,38 +190,54 @@ func (c *controllerCommon) AttachDisk(ctx context.Context, diskName, diskURI str
 		return -1, err
 	}
 
-	var waitForDetachHappened bool
-	if c.WaitForDetach {
-		// wait for disk detach to finish first on the same node
-		if err = wait.PollUntilContextTimeout(ctx, 2*time.Second, 30*time.Second, true, func(context.Context) (bool, error) {
-			detachDiskReqeustNum, err := c.getDetachDiskRequestNum(node)
-			if err != nil {
-				return false, err
-			}
-			if detachDiskReqeustNum == 0 {
-				return true, nil
-			}
-			klog.V(4).Infof("there are still %d detach disk requests on node %s, wait for detach to finish, current disk: %s", detachDiskReqeustNum, node, diskName)
-			waitForDetachHappened = true
-			return false, nil
-		}); err != nil {
-			klog.Errorf("current disk: %s, wait for detach disk requests on node %s failed: %v", diskName, node, err)
-		}
-	}
-
 	c.lockMap.LockEntry(node)
 	defer c.lockMap.UnlockEntry(node)
 
-	if !waitForDetachHappened && c.AttachDetachInitialDelayInMs > 0 && requestNum == 1 {
+	if c.AttachDetachInitialDelayInMs > 0 && requestNum == 1 {
 		klog.V(2).Infof("wait %dms for more requests on node %s, current disk attach: %s", c.AttachDetachInitialDelayInMs, node, diskURI)
 		time.Sleep(time.Duration(c.AttachDetachInitialDelayInMs) * time.Millisecond)
 	}
 
+	numDisksAllowed := math.MaxInt
+	if c.CheckDiskCountForBatching {
+		_, instanceType, err := getNodeInfoFromLabels(ctx, string(nodeName), c.cloud.KubeClient)
+		if err != nil {
+			return -1, err
+		}
+		if instanceType != "" {
+			maxNumDisks, instanceExists := getMaxDataDiskCount(instanceType)
+			if instanceExists {
+				attachedDisks, _, err := c.GetNodeDataDisks(ctx, nodeName, azcache.CacheReadTypeDefault)
+				if err != nil {
+					return -1, err
+				}
+				numDisksAttached := len(attachedDisks)
+				numDisksAllowed = int(maxNumDisks) - numDisksAttached
+			}
+		}
+	}
+
 	diskMap, err := c.cleanAttachDiskRequests(node)
 	if err != nil {
 		return -1, err
 	}
 
+	// Remove some disks from the batch if the number is more than the max number of disks allowed
+	removeDisks := len(diskMap) - numDisksAllowed
+	if removeDisks > 0 {
+		klog.V(2).Infof("too many disks to attach, remove %d disks from the request", removeDisks)
+		for diskURI, options := range diskMap {
+			if removeDisks == 0 {
+				break
+			}
+			if options != nil {
+				klog.V(2).Infof("remove disk(%s) from attach request", diskURI)
+				delete(diskMap, diskURI)
```

Comment:
If a disk is removed from the map, the error that will be returned is "disk(<pvc>) could not be found on node" which is confusing. We should emit a clearer error message, e.g. "disk <pvc> could not be attached to node <node> because it would exceed the maximum number of disks".

Reply:
this kind of clear error msg may be difficult to set since in this AttachDisk operation, it would return success since other disks are removed from the list.

Reply:
The disk that is removed may have come from a different attach request, so we'd need to find a way to let that request know that the disk from another batch attach request has dropped the disk. 

One way to do that would be to cache all the removed disks in a map and if we find the disk from the current request in that map, we return the appropriate error. But that would require maintaining another map and using locks for concurrency control. Please lmk if you have any better ideas.

Reply:
that's too complex and error prone, I think we could just keep current error msg if there is no better way, it's nice to have anyway.